File: backend/app/limiters/fixed_window.py
from __future__ import annotations

from redis.asyncio import Redis

from app.limiters.base import Decision


class FixedWindowLimiter:
    algorithm = "fixed_window"

    async def allow(
        self,
        redis_client: Redis,
        policy_id: str,
        client_id: str,
        now_ms: int,
        params: dict[str, object],
    ) -> Decision:
        window_size_sec = int(params["window_size_sec"])
        limit = int(params["limit"])

        window_ms = window_size_sec * 1000
        window_start_ms = (now_ms // window_ms) * window_ms
        key = f"rl:fixed_window:{policy_id}:{client_id}:{window_start_ms}"

        count = int(await redis_client.incr(key))
        if count == 1:
            await redis_client.expire(key, window_size_sec + 1)

        if count <= limit:
            return Decision(
                allowed=True,
                remaining=max(limit - count, 0),
                algorithm_state={"count": count, "window_start_ms": window_start_ms},
            )

        retry_after_ms = max(window_start_ms + window_ms - now_ms, 1)
        return Decision(
            allowed=False,
            reason="rate_limit_exceeded",
            remaining=0,
            retry_after_ms=retry_after_ms,
            algorithm_state={"count": count, "window_start_ms": window_start_ms},
        )

File: gateway/internal/limiter/fixed_window.go
package limiter

import (
	"context"
	"fmt"
)

type FixedWindowLimiter struct {
	store RuntimeCounter
}

func NewFixedWindowLimiter(store RuntimeCounter) *FixedWindowLimiter {
	return &FixedWindowLimiter{store: store}
}

func (l *FixedWindowLimiter) Allow(ctx context.Context, policyID, clientID string, nowMS int64, params map[string]any) (Decision, error) {
	windowSizeSec, err := readPositiveInt(params, "window_size_sec")
	if err != nil {
		return Decision{}, err
	}
	limit, err := readPositiveInt(params, "limit")
	if err != nil {
		return Decision{}, err
	}

	windowMS := int64(windowSizeSec * 1000)
	windowStartMS := (nowMS / windowMS) * windowMS
	key := fmt.Sprintf("rl:fixed_window:%s:%s:%d", policyID, clientID, windowStartMS)

	count, err := l.store.Incr(ctx, key)
	if err != nil {
		return Decision{}, err
	}
	if count == 1 {
		if err := l.store.Expire(ctx, key, windowSizeSec+1); err != nil {
			return Decision{}, err
		}
	}

	remaining := limit - int(count)
	if count <= int64(limit) {
		if remaining < 0 {
			remaining = 0
		}
		return Decision{
			Allowed:   true,
			Remaining: &remaining,
			AlgorithmState: map[string]any{
				"count":           count,
				"window_start_ms": windowStartMS,
			},
		}, nil
	}

	retryAfter := int((windowStartMS + windowMS) - nowMS)
	if retryAfter < 1 {
		retryAfter = 1
	}
	reason := "rate_limit_exceeded"
	zero := 0
	return Decision{
		Allowed:      false,
		Reason:       &reason,
		Remaining:    &zero,
		RetryAfterMS: &retryAfter,
		AlgorithmState: map[string]any{
			"count":           count,
			"window_start_ms": windowStartMS,
		},
	}, nil
}

func readPositiveInt(input map[string]any, key string) (int, error) {
	value, ok := input[key]
	if !ok {
		return 0, fmt.Errorf("missing required field %s", key)
	}
	switch v := value.(type) {
	case float64:
		if v <= 0 {
			return 0, fmt.Errorf("field %s must be > 0", key)
		}
		return int(v), nil
	case int:
		if v <= 0 {
			return 0, fmt.Errorf("field %s must be > 0", key)
		}
		return v, nil
	default:
		return 0, fmt.Errorf("field %s must be numeric", key)
	}
}
